specs/Print.md

```diff
@@ -0,0 +1,816 @@
+
+**Print to PDF Stream**: This API gives developers printable PDF data, which can be used as a preview
+in a custom print preview dialog or print pdf data of current web page using native printer APIs.
+This API consists of an asynchronous PrintToPdfStream method and a PrintSettings2 object.
+
+
+# Examples
+## ShowPrintUI
+
+You can use the `ShowPrintUI` method to programmatically open a print dialog. Currently devs are
+using JavaScript `window.print()` using the `CoreWebView2.ExecuteScriptAsync` API to invoke the
+browser print preview dialog. But there isn't a way to invoke the system print dialog
+programmatically nor will `window.print()` work if script is disabled on the page, or the
+document is a PDF or other non-HTML document.
+
+```cpp
+// This example shows the user a print dialog. If `printDialogKind` is
+// COREWEBVIEW2_PRINT_DIALOG_KIND_BROWSER, opens a browser print preview dialog,
+// otherwise opens a system print dialog.
+
+bool AppWindow::ShowPrintUI(COREWEBVIEW2_PRINT_DIALOG_KIND printDialogKind)
+{
+  wil::com_ptr<ICoreWebView2_15> webView2_15;
+  CHECK_FAILURE(m_webView->QueryInterface(IID_PPV_ARGS(&webView2_15)));
+  CHECK_FEATURE_RETURN(webView2_15);
+  CHECK_FAILURE(webView2_15->ShowPrintUI(printDialogKind));
+  return true;
+}
+```
+
+```c#
+void ShowPrintUI(object target, ExecutedRoutedEventArgs e)
+{
+  string printDialog = e.Parameter.ToString();
+  if (printDialog == "Browser")
+  {
+    // Opens the browser print preview dialog.
+    webView.CoreWebView2.ShowPrintUI();
+  }
+  else
+  {
+    // Opens the system print dialog.
+    webView.CoreWebView2.ShowPrintUI(CoreWebView2PrintDialogKind.System);
+  }
+}
+```
+
+# Print to default printer
+
+You can use `Print` method to print to a default printer without a print dialog.
+You can either use default page and printer settings or programmatically configure
+print settings, and call `Print` to print the current web page.
+
+```cpp
+// This example prints the current web page without a print dialog to a default printer
+// with the default print settings.
+
+bool AppWindow::PrintToDefaultPrinter()
+{
+  wil::com_ptr<ICoreWebView2_15> webView2_15;
+  CHECK_FAILURE(m_webView->QueryInterface(IID_PPV_ARGS(&webView2_15)));
+
+  wil::com_ptr<ICoreWebView2Environment6> webviewEnvironment6;
+  CHECK_FAILURE(m_webViewEnvironment->QueryInterface(IID_PPV_ARGS(&webviewEnvironment6)));
+
+  wil::com_ptr<ICoreWebView2PrintSettings> printSettings = nullptr;
+  CHECK_FAILURE(webviewEnvironment6->CreatePrintSettings(&printSettings));
+
+  wil::com_ptr<ICoreWebView2PrintSettings2> printSettings2;
+  CHECK_FAILURE(printSettings->QueryInterface(IID_PPV_ARGS(&printSettings2)));
+
+  wil::unique_cotaskmem_string title;
+  CHECK_FAILURE(m_webView->get_DocumentTitle(&title));
+
+  // Prints current web page with the default page and printer settings.
+  CHECK_FAILURE(webView2_15->Print(
+      printSettings2.get(),
```

Comment:
I think the Print method's first parameter should be a plain old `ICoreWebView2PrintSettings` (not the `2` version). That would also remove the need for QI for `ICoreWebView2PrintSettings2` at line 101.

Reply:
Right, passing a `ICoreWebView2PrintSettings2` is acceptable for functions that accept a `ICoreWebView2PrintSettings`. So just accepting a `ICoreWebView2PrintSettings` is the most accommodating parameter type. And it makes sense, because maybe you're only customizing the v1 properties, so you never needed to query for `ICoreWebView2PrintSettings2` in the first place.